pallets/public-credentials/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+		// Vec<u8> instead of BoundedVec<u8, ...> because otherwise it becomes
+		// impossible for runtime-common to be independent of the `T: Config`
+		// constraint.
+		type SubjectId: Parameter + MaxEncodedLen + TryFrom<Vec<u8>>;
+		/// The weight info.
+		type WeightInfo: WeightInfo;
+
+		/// The amount of tokens to reserve when attesting a public credential.
+		#[pallet::constant]
+		type Deposit: Get<BalanceOf<Self>>;
+		/// The maximum length in bytes of the encoded claims of a credential.
+		#[pallet::constant]
+		type MaxEncodedClaimsLength: Get<u32>;
+		/// The maximum length in bytes of the raw credential subject
+		/// identifier.
+		#[pallet::constant]
+		type MaxSubjectIdLength: Get<u32>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	/// The map of public credentials already attested.
+	/// It maps from a (subject id + credential id) -> the creation
+	/// details of the credential.
+	#[pallet::storage]
+	#[pallet::getter(fn get_credential_info)]
+	pub type Credentials<T> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		<T as Config>::SubjectId,
+		Blake2_128Concat,
+		CredentialIdOf<T>,
+		CredentialEntryOf<T>,
+	>;
+
+	/// A reverse index mapping from credential ID to the subject the credential
+	/// was issued to.
+	///
+	/// It it used to perform efficient lookup of credential given their ID.
+	#[pallet::storage]
+	#[pallet::getter(fn get_credential_subject)]
+	pub type CredentialSubjects<T> = StorageMap<_, Blake2_128Concat, CredentialIdOf<T>, <T as Config>::SubjectId>;
+
+	/// The events generated by this pallet.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A new public credential has been issued.
+		CredentialStored {
+			/// The subject of the new credential.
+			subject_id: T::SubjectId,
+			/// The id of the new credential.
+			credential_id: CredentialIdOf<T>,
+		},
+		/// A public credentials has been removed.
+		CredentialRemoved {
+			/// The subject of the removed credential.
+			subject_id: T::SubjectId,
+			/// The id of the removed credential.
+			credential_id: CredentialIdOf<T>,
+		},
+		/// A public credential has been revoked.
+		CredentialRevoked {
+			/// The id of the revoked credential.
+			credential_id: CredentialIdOf<T>,
+		},
+		/// A public credential has been unrevoked.
+		CredentialUnrevoked {
+			/// The id of the unrevoked credential.
+			credential_id: CredentialIdOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// A credential with the same root hash has already issued to the
+		/// specified subject.
+		CredentialAlreadyIssued,
+		/// No credential with the specified root hash has been issued to the
+		/// specified subject.
+		CredentialNotFound,
+		/// Not enough tokens to pay for the fees or the deposit.
+		UnableToPayFees,
+		/// The credential input is invalid.
+		InvalidInput,
```

Comment:
This error is not tested!

Reply:
Good catch! https://github.com/KILTprotocol/kilt-node/pull/378/commits/51447051f4ccc1c953316988a7f89ad6256904de